pyomo/repn/plugins/nl_writer.py

```diff
@@ -710,6 +704,74 @@ def write(self, model):
         else:
             timer.toc('Processed %s constraints', len(all_constraints))
 
+        # Complementarity handling
+        n_complementarity_nonlin = 0
+        n_complementarity_lin = 0
+        n_complementarity_range = 0
+        n_complementarity_nz_var_lb = 0
+
+        for block in component_map[Complementarity]:
+            for comp in block.component_data_objects(
+                Complementarity, active=True, descend_into=False, sort=sorter
+            ):
+                # Transform the complementarity condition into standard form
+                comp.to_standard_form()
+
+                # If a new variable was created, add it to var_map
+                if hasattr(comp, 'v'):
+                    _id = id(comp.v)
+                    if _id not in var_map:
+                        var_map[_id] = comp.v
+
+                    # Check if variable has nonzero lower bound or any upper bound
+                    lb, ub = comp.v.bounds
+                    if (lb is not None and lb != 0) or ub is not None:
+                        n_complementarity_nz_var_lb += 1
+
+                # Process the complementarity constraint
+                if hasattr(comp, 'c'):
+                    con = comp.c
+                    con._vid = _id
+                    if hasattr(con, '_complementarity_type'):
+                        con._complementarity = con._complementarity_type
```

Comment:
I see that the  _complementarity attribute is currently used below for the 'r' lines writing bit.
_complementarity is mainly used in mpec4.py and ampl_.py while _complementarity is used in complementarity.py and mpec1.py. Are they not the same ? If they are the same, I am happy to raise a separate PR to rename one into the other